### PR TITLE
optimize calculation of input jacobian

### PR DESCRIPTION
**Pull request by baggepinnen**
This PR builds upon
- #1680

### CODE REVIEW

File: src/systems/abstractsystem.jl [hunk 1]
```diff
@@ -1040,8 +1044,19 @@ function linearization_function(sys::AbstractSystem, inputs,
                 fg_xz = ForwardDiff.jacobian(uf, u)
                 h_xz = ForwardDiff.jacobian(xz -> h(xz, p, t), u)
                 pf = SciMLBase.ParamJacobianWrapper(fun, t, u)
-                # TODO: this is very inefficient, p contains all parameters of the system
-                fg_u = ForwardDiff.jacobian(pf, p)[:, input_idxs]
+
+                p_closure = function (p_small_inner)
```
Comment (YingboMa) [inline src/systems/abstractsystem.jl:1048]: You need a `let` to make sure that it's type stable.

**Comment by baggepinnen** [comment]
I don't get why the format check is failing, I get
```julia
julia> using JuliaFormatter, ModelingToolkit; format(joinpath(dirname(pathof(ModelingToolkit)), ".."))
true
```
locally.

File: src/systems/abstractsystem.jl [hunk 2]
```diff
@@ -1040,8 +1044,21 @@ function linearization_function(sys::AbstractSystem, inputs,
                 fg_xz = ForwardDiff.jacobian(uf, u)
                 h_xz = ForwardDiff.jacobian(xz -> h(xz, p, t), u)
                 pf = SciMLBase.ParamJacobianWrapper(fun, t, u)
-                # TODO: this is very inefficient, p contains all parameters of the system
-                fg_u = ForwardDiff.jacobian(pf, p)[:, input_idxs]
+
+                p_closure = let pf = pf
+                    function (p_small_inner::T) where {T}
+                        p_big[input_idxs] .= p_small_inner
+                        pf(p_big)::T
```
Comment (YingboMa) [inline src/systems/abstractsystem.jl:1051]: Is this true? The parameters could be `Vector{Int}` while the observed variables are `Vector{Float64}`

File: src/systems/abstractsystem.jl [hunk 3]
```diff
@@ -1040,8 +1044,21 @@ function linearization_function(sys::AbstractSystem, inputs,
                 fg_xz = ForwardDiff.jacobian(uf, u)
                 h_xz = ForwardDiff.jacobian(xz -> h(xz, p, t), u)
                 pf = SciMLBase.ParamJacobianWrapper(fun, t, u)
-                # TODO: this is very inefficient, p contains all parameters of the system
-                fg_u = ForwardDiff.jacobian(pf, p)[:, input_idxs]
+
+                p_closure = let pf = pf
```
Comment (YingboMa) [inline src/systems/abstractsystem.jl:1048]: `p_big` and `input_idxs` are also enclosed.

**Review by YingboMa — COMMENTED** [review]
I think we should add this functionality in SparseDiffTools.jl

**Comment by YingboMa** [comment]
Superseded by #1724

**Pull request closed without merging**